### PR TITLE
Optimize cpp reader

### PR DESCRIPTION
**Pull request by jacquesqiao**
### speed test
| reader| speed (sample/second) | 
| ------ | ------ |
| fake py_reader| 18273.2305371 | 
| ctr_reader thread = 1 |37429.8865173 |
| ctr_reader thread = 2 |75221.685231 |
| ctr_reader thread = 4 |142344.240397 |
| ctr_reader thread = 8 |246912.289501 |

### use in [ctr task](https://github.com/PaddlePaddle/models/tree/develop/fluid/PaddleRec/ctr)
| batch_size| Fake pyreader (sample/second) | test_mode (sample/second) | ctr_reader (sample/second)
| ------ | ------ | ------ | ---- |
| 128 | 18273.2305371 | 31472.6573437 |28486.2236257
| 1024 |19085.3616117 |60088.7414079| 55692.4481658

### CODE REVIEW

**Review by typhoonzero — DISMISSED** [review]
Only reviewed API change!

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 1]
```diff
@@ -44,13 +50,18 @@ def _copy_reader_var_(block, var):
...
 
-def ctr_reader(feed_data,
-               capacity,
-               thread_num,
-               batch_size,
-               file_list,
-               slots,
-               name=None):
+def ctr_reader(
+        feed_dict,
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py:54]: doc?

**Comment by jacquesqiao** [comment]
merged by other pr

File: python/paddle/fluid/contrib/reader/ctr_reader.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/fluid/contrib/reader/ctr_reader.py:54]: done

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 2]
```diff
@@ -67,11 +78,20 @@ def ctr_reader(feed_data,
...
+       file_type('gzip'|'plain'): the type of the data file
+       file_format('csv'|'svm'): csv data or svm data format.
+        cvs data format is :
+            label dense_fea,dense_fea sparse_fea,sparse_fea
+        the svm data format is :
+            label slot1:fea_sign slot2:fea_sign slot1:fea_sign
+       dense_slot_index(list(int)): the index of dense slots
+       sparse_slot_index(list(int)): the index of sparse slots
        capacity(int): The buffer capacity maintained by :code:`py_reader`.
        thread_num(list|tuple): List of tuples which declaring data shapes.
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py]: ?

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 3]
```diff
@@ -67,11 +78,20 @@ def ctr_reader(feed_data,
...
        batch_size(list|tuple): List of strs which declaring data type.
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py]: ?

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 4]
```diff
@@ -67,11 +78,20 @@ def ctr_reader(feed_data,
...
        file_list(list|tuple): List of ints which declaring data lod_level.
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py]: ?

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 5]
```diff
@@ -67,11 +78,20 @@ def ctr_reader(feed_data,
...
-       slots(bool): Whether use double buffer or not.
+       slots(bool): slot id of all sparse feature
        name(basestring): The prefix Python queue name and Reader name. None will
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py]: Examples is empty

Comment (panyx0718) [reply python/paddle/fluid/contrib/reader/ctr_reader.py]: need data format example
api input example

File: python/paddle/fluid/contrib/reader/ctr_reader.py [hunk 6]
```diff
@@ -67,11 +78,20 @@ def ctr_reader(feed_data,
     Note that :code:`Program.clone()` method cannot clone :code:`py_reader`.
 
     Args:
+       feed_dict(list(variable)): a list of data variable.
```
Comment (panyx0718) [inline python/paddle/fluid/contrib/reader/ctr_reader.py:81]: so it's not list? update comment?

Comment (jacquesqiao) [reply python/paddle/fluid/contrib/reader/ctr_reader.py:81]: It is a list of variable

File: python/paddle/fluid/contrib/reader/ctr_reader.py (hunk 5 above)
Comment (jacquesqiao) [reply python/paddle/fluid/contrib/reader/ctr_reader.py]: add example

File: python/paddle/fluid/contrib/reader/ctr_reader.py (hunk 4 above)
Comment (jacquesqiao) [reply python/paddle/fluid/contrib/reader/ctr_reader.py]: fixed

**Review by panyx0718 — APPROVED** [review]

**Comment by panyx0718** [comment]
This API will be hidden by the pyreader by Q1

**Pull request merged**